



# 同步与异步

同步: 等待结果

异步: 不等待结果

从`js`角度去看:

- `js`是单线程的  

  - 单线程基本上是无法等待的
  - 如果单线程还等待的话,剩下的工作由谁去完成呢?
  - 所以js的特性,导致js的代码都是同步代码,也就是不会等待的代码

- 异步

  - 下一句的代码要等待上一句的结果,才可以进行,这就是异步
  - `js`表现出来的异步,其实是事情交给浏览器去处理,js会继续执行下一句代码
  - 所以的DOM事件都是异步的
    - 我们写js代码实现事件的绑定,然后js会接着执行下面的代码
    - 浏览器负责监听事件,一旦事件触发,浏览器会告诉js引擎去执行事件的回调函数

  - 异步常常伴随着回调一起出现,但这两个没有决定的关系
    - 出现这一现象的原因,是浏览器的机制导致的,就是上面我说的那个过程

  

  

# 前端经常遇到的异步

```javascript
document.getElementsByTagNames("img")[0] .width // 宽度为 0
  console•log('done')
```

上述代码看起来是同步的,但是实际上是异步的!

- 创建了一个DOM节点,对应页面上的`img`元素,之后获取元素的width
- 但是下载 `img`的是浏览器,而浏览器的下载需要时间,导致出现问题
  - 浏览器根据`img`的`src`发送请求
  - 服务器接收请求,向浏览器响应数据
  - 浏览器接收响应,下载数据
  - 渲染页面
- `js`的同步代码,导致width为0 
- 注意:上述的问题只会在第一次页面第一次执行时出现
  - 原因: 浏览器的缓存
  - 得到数据之后,再次运行`js` ,也就没有异步了 (不需要发送请求了!)

**解决方案:**

获取`img`的width这个操作,作为onload事件的回调

```javascript
document.getElementsByTagNames('img*)[0].onload = function(){
         console.log(this.width) // 室度不为 0
         console.log('real done')

}
```



**反思:**

==凡是需要需要发送请求的代码都是异步的!==

因为都是浏览器发送请求,下载得到的数据!  这个过程是需要时间的 ! 而`js`的代码是不会等待的 !   

这种情况,把**请求**作为事件的回调函数去写代码,回避这个问题!

[img  style script  a  link  这些发送请求的方式中,最容易犯错的确实就是img]()



# 面试中经常遇到的异步

```javascript
let liList = document.querySelectAll('li')
for(var i=0;i<liList.length;i++){
    liList[i].onclick = function(){
        console.log(i)
    }
}
// 绑定的结果是点击任何一个li,打印的结果都是 6
```

这个问题想要彻底理解需要明白两个关键:

- for()的本质
- 异步

```javascript
for(var i=0;i<liList.length;i++){
    liList[i].onclick = function(){
        console.log(i)
    }
==>
i=0  liList[i].onclick = function(){
        console.log(i)
    } *6  每次的i++,其他的代码不变

// 1. 上面的代码,相当于做了6次事件绑定
// 2. js的不等待原则,一口气执行了所有的代码 , 代码运行结束时 函数的环境: i=6
// 3. 当事件触发,js执行回调时,i已经是6了
```

异步

​	`js`依次执行上述代码,没事件触发,所以只执行事件的绑定 ,  监听就交给浏览器了  

​	浏览器监听到事件触发告诉`js`回调时,此时全局里面的`i-->6`了

​	造成了上面的问题

解决方案:

- 把`var i ==> let i`  就解决了
- 主要let的作用域很好用  我在ES6深入浅出,详细介绍了
- 不懂的就去看看
- 好吧,我知道你肯定懒得去看! 我就简单说说!
  - `console.log(i)`每次回调执行时, 总是6  
  - 因为 这个回调函数运行时需要依赖函数外部的变量   (词法分析可知道,[你真的懂函数吗](C:\Users\lenovo\Documents\一"桶"前端\myNote\js深入浅出\01-你真的懂函数吗.md))
  - 你可能会疑惑:为什么回调函数是依赖外部的变量?
    - for()循环是写在全局里面的 , 里面`var i`也是在全局里面的
    - 执行回调时要用到 `i ` , 也就依赖了外部的变量
    - 用`let i`之后 ,  `i`就是函数内部的变量(`js的用let引入了局部作用域`)(具体怎么实现的,不需要知道,只要己住就好了)



# AJAX 中的异步

```javascript
 let request = $.ajax({
      url:
      async: false

   })
   console.log(request.responseText)
```

这就是后端代码者的思维:

主要原因就是 java等后端语言都是多线程的,所以代码是可以等待的,下面的代码不会阻塞,会有其他的线程去执行!

但是上述代码放到js中就不行了!!! 

js不会去执行后面的代码,会一直执行查询代码,查看请求是否成功了,去获取响应的数据

**解决方案:**

==依然是利用事情机制,让浏览器监听,然后通知js执行回调函数==

```javascript
$.ajax({
         url:
         async: true,
    //用一个succes选项,轻松解决异步的问题
         success: function(responseText){
               console.log(responseText)

   })
```





# 重要思考(自己瞎想的)

java等后端语言解决代码异步 是通过多线程解决的

js解决方式: 利用浏览器的监听事件的机制,通知js回调,解决代码异步





# 如何拿到异步结果

一般有两种方式:

- 傻逼方法: 轮询  (早就不用了)
- 正规方法: 回调

**如何知道异步操作的结果是成功的呢还是失败的呢?**

==还是通过回调!==

## 回调的形式

1. Nodejs 的 error-first 形 式

```javascript

fs.readFile('./l.txt', (error, content)=>{
      if(error){
      //失畋
}else{
        //成功
     }
})
```

2. jQuery 的 success / error 形 式

```javascript
 $.ajax({
             url: Vxxx',
             success:()=>{},
             error: ()=>{}
       })

```

3. jQuery 的 done / fail / always 形 式

```javascript
   $.ajax({
              url:'/xxx1,
        }).done( ()=>{} ).fail( ()=>{} ).always( ()=> {})
// 这个暂且看不懂 , 要理解函数的柯里化才可以
```

4. `Prosmise `的 then 形 式

```javascript
 $.ajax({
       url: Vxxx*,
 }).then( ()=>{}, ()=>{} ).then( ()=>{})
```



## 小结

可以看到 判断异步执行是否成功的方式五花八门

如: Node.js中 通过判断执行回调函数的传参个数来知道异步是否成功

​	jQuery 中通过 success和erorr两个选项分别对应不同的回调,来判断异步结果

最终AJAX这个著名的异步操作,通过 Promise 统一了 判断异步操作是否成功的格式!





# Promise

源码分析

```javascript
class Promise{
  constructor(executor){
    //控制状态，使用了一次之后，接下来的都不被使用
    this.status = 'pendding'
    this.value = undefined
    this.reason = undefined
    
    //定义resolve函数
    let resolve = (data)=>{
      //这里pendding，主要是为了防止executor中调用了两次resovle或reject方法，而我们只调用一次
      if(this.status==='pendding'){
        this.status = 'resolve'
        this.value = data
      } 
    }

    //定义reject函数
    let reject = (data)=>{
      if(this.status==='pendding'){
        this.status = 'reject'        
        this.reason = data
      } 
    }

    //executor方法可能会抛出异常，需要捕获
    try{
      //将resolve和reject函数给使用者      
      executor(resolve,reject)      
    }catch(e){
      //如果在函数中抛出异常则将它注入reject中
      reject(e)
    }
  }
}

```

